syntax = "proto3";
package trading;

service TradingEngine {
	rpc AddOrder(OrderRequest) returns (TradeResponse);
	rpc CancelOrder(CancelOrderRequest) returns (CancelOrderResponse);
	rpc ModifyOrder(ModifyOrderRequest) returns (TradeResponse);
	rpc GetOrderbook(OrderbookRequest) returns (OrderbookResponse);
}

enum OrderType {
	ORDER_TYPE_UNSPECIFIED = 0;
	GOOD_TILL_CANCEL = 1;
	FILL_AND_KILL = 2;
	FILL_OR_KILL = 3;
	GOOD_FOR_DAY = 4;
	MARKET = 5;
}

enum Side {
	SIDE_UNSPECIFIED = 0;
	BUY = 1;
	SELL = 2;
}

enum OrderStatus {
	ORDER_STATUS_UNSPECIFIED = 0;
	ACCEPTED = 1;
	REJECTED = 2;
	FILLED = 3;
	PARTIALLY_FILLED = 4;
	CANCELLED = 5;
}

message OrderRequest {
	uint64 order_id = 1;
	Side side = 2;
	int32 price = 3;
	uint32 quantity = 4;
	OrderType order_type = 5;
}

message TradeInfo {
	uint64 order_id = 1;
	int32 price = 2;
	uint32 quantity = 3;
}

message TradeResponse {
	uint64 order_id = 1;
	OrderStatus status = 2;
	repeated TradeInfo trades = 3;
}

message CancelOrderRequest {
	uint64 order_id = 1;
}

message CancelOrderResponse {
	bool success = 1;
}

message ModifyOrderRequest {
	uint64 order_id = 1;
	Side side = 2;
	int32 new_price = 3;
	uint32 new_quantity = 4;
}

// Minimal orderbook response
message LevelInfo {
	int32 price = 1;
	uint32 quantity = 2;
}

message OrderbookRequest {
	// Empty - single instrument for now
}

message OrderbookResponse {
	repeated LevelInfo bids = 1;
	repeated LevelInfo asks = 2;
}
